rust/arrow/src/array/array.rs

```diff
@@ -2319,13 +2221,24 @@ impl<T: ArrowPrimitiveType> From<ArrayDataRef> for DictionaryArray<T> {
             "DictionaryArray should contain a single child array (values)."
         );
 
-        let raw_values = data.buffers()[0].raw_data();
-        let dtype: &DataType = data.data_type();
-        let values = make_array(data.child_data()[0].clone());
-        if let DataType::Dictionary(_, _) = dtype {
+        if let DataType::Dictionary(key_data_type, _) = data.data_type() {
+            if key_data_type.as_ref() != &T::DATA_TYPE {
+                panic!("DictionaryArray's data type must match.")
```

Comment:
```suggestion
                unreachable!("DictionaryArray's data type must match.")
```
Since the former is good for defensive programming but doesn't convey the idea.

Reply:
Thanks for the suggestions.

Isn't `unreachable` used when the program arrives at an inconsistent state? 

IMO, in this case, we are checking user input (this function is public) and ensure that we will not reach an inconsistent state (in the same way `assert_eq` does). `assert_eq` calls `panic!`, which is why I also used `panic!` here.

Reply:
I don't think that unreachable is only for the inconsistent state. But we can leave it as panic here. I was also unsure about the user-facing API's panicking behavior. Especially in array methods with forced asserts. We should prefer Result than direct asserts, but you know... That's also yet another topic.